package.json

```diff
@@ -70,6 +70,7 @@
     "patch-package": "^6.2.2"
   },
   "resolutions": {
+    "**/ast-types": "github:agoric-labs/ast-types#Agoric-built",
```

Comment:
I'd like to understand why this is necessary now, and wasn't necessary before?

Reply:
Curious as well. `ast-types` is brought in via `recast` via `@endo/static-module-record` via `@endo/compartment-mapper` via bundle utilities. We had to modify it to make it work under SES. It can’t be a patch package because the patch would need to be distributed to dapps and any other dependee transitively. We may need to also patch `ast-types` so it runs under `-r esm` if we end up spending a lot of time supporting contract authors who stub their toe on `ast-types does not have a default export` errors often enough or if rolling forward is hard enough.

There shouldn’t be a need for `ast-types` to be deduplicated. Perhaps there’s another dependency path that brings in the wrong `ast-types` and `yarn` is deduping them to our detriment?

Reply:
I'd also like to know! This resolution is just to make my laptop write out the same yarn.lock that CI does. Incidentally, why do we have a `git diff` check instead of `yarn install --frozen-lockfile`? 

Thanks for pointing out why it can't be a patch. How about a versioned release? 

> There shouldn’t be a need for ast-types to be deduplicated. Perhaps there’s another dependency path that brings in the wrong ast-types and yarn is deduping them to our detriment?

That would make sense, but there's just one thing depending on `ast-types`:
```
recast@agoric-labs/recast#Agoric-built:
  version "0.20.5"
  resolved "https://codeload.github.com/agoric-labs/recast/tar.gz/879398a55cd50a53ade179de203706a25c53fb49"
  dependencies:
    ast-types "github:agoric-labs/ast-types#Agoric-built"
    esprima "~4.0.0"
    source-map "~0.6.1"
    tslib "^2.0.1"
```

Reply:
> How about a versioned release?

That was suggested be @kriskowal as an alternative. Given the release complexity we ran into with github, and now this, we should seriously consider this again.



> Incidentally, why do we have a `git diff` check instead of `yarn install --frozen-lockfile`?

Apparently that doesn't result in a non-zero exit code if it fails...

Reply:
> Apparently that doesn't result in a non-zero exit code if it fails...

I've heard rumors of this but not witnessed it. [The docs](https://classic.yarnpkg.com/lang/en/docs/cli/install/#toc-yarn-install-frozen-lockfile) say it does and a test locally does,
<img width="753" alt="Screen Shot 2022-02-23 at 4 03 30 PM" src="https://user-images.githubusercontent.com/21505/155431618-1a2d9b39-49a7-4166-a2c1-9a1677441cac.png">

It's also what I used for the past several years without incident.

Reply:
I don't object to your questioning my judgement.  But I don't want to be the one to have to repeat diagnosing the problem I ran into where `yarn --frozen-lockfile` didn't fail with an error but the lockfile still differed.

Reply:
Nothing personal here. My last comment  above was in response to the assertion "doesn't result in a non-zero exit code if it fails". 

My questioning whether we really need to have our own diff check instead of `frozen-lock` preceded your answer in chat,
> In the testing I did around this, the changes I was able to induce without --frozen-lockfile failing were (material) semantic ones.

That information leads me to a different questions:
- what is a way to reproduce that bug in frozen-lockfile? Not because I doubt that you encountered it, but because 1) it might have been fixed in yarn since then and 2) if it hasn't I'd like to file it as a bug
- whether there are downsides to using `--frozen-lockfile` anyway, to convey that intent (keeping the `if git is dirty` as a backup for this bug)

In the interests of reducing work in flight I'll make that change optimistically and request a code review.

Reply:
> Nothing personal here.

No worries!  I was being unnecessarily short.

> My last comment above was in response to the assertion "doesn't result in a non-zero exit code if it fails".

Steps to reproduce:

1. delete, say, the `@agoric/nat^4.1.0` section from `yarn.lock`.
2. run `yarn install --frozen-lockfile`.  It doesn't put it back (check the `git diff`), but it also doesn't error.
3. run `yarn install`, it puts the `@agoric/nat^4.1.0` section back

So the issue I found was that `--frozen-lockfile` doesn't complain about the things that a normal `yarn install` would have changed.  I'm more interested in making sure that `yarn install` wouldn't have changed anything, not just that the `yarn.lock` isn't modified.